<a href="https://colab.research.google.com/github/coolmino96/Predicting-MPG/blob/master/%EA%B3%BC%EC%A0%9C2_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import tensorflow as tf

In [9]:
tf.__version__ 

'2.3.0'

In [10]:
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 0s 0us/step


**1. Layer의 갯수 1개 vs 3개**

In [27]:
model = tf.keras.models.Sequential([
                                    
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [28]:
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3596 - accuracy: 0.8975
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2990 - accuracy: 0.9160
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2868 - accuracy: 0.9204
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2822 - accuracy: 0.9206
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2765 - accuracy: 0.9231
313/313 - 0s - loss: 0.2729 - accuracy: 0.9272


[0.2729000747203827, 0.9272000193595886]

In [29]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3916 - accuracy: 0.8867
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3345 - accuracy: 0.9044
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3192 - accuracy: 0.9087
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3115 - accuracy: 0.9122
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3017 - accuracy: 0.9140
313/313 - 1s - loss: 0.3369 - accuracy: 0.9053


[0.3369215726852417, 0.9053000211715698]

*   layer를 늘리면 당연히 결과가 좋을 줄 알았는데 아닙니다.
*   결과를 보면 사실 크게 차이가 없다는 것을 알 수 있습니다.
*   train set보다 test set의 loss가 더 낮고, accuracy도 더 높습니다. 일종의 underfitting 이라고 볼 수 있습니다.
*   일반화 성능보다, 일단 train set 자체를 더 잘 학습하는 게 중요하며, 위 결과처럼 layer만 늘린다고 성능이 좋지는 않습니다
*   딥러닝의 Vanishing Gradients 문제로 인해서 layer를 늘린만큼 학습이 잘 되도록 몇가지 변화를 주어야 합니다.

**2. ReLU의 유무**

In [30]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2239 - accuracy: 0.9333
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0999 - accuracy: 0.9688
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0716 - accuracy: 0.9784
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0551 - accuracy: 0.9831
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0467 - accuracy: 0.9855
313/313 - 1s - loss: 0.0796 - accuracy: 0.9755


[0.07956139743328094, 0.9754999876022339]



*   Train set의 학습자체도 빠르고 성능이 좋습니다.
*   Test set의 결과도 아주 좋아졌습니다.



**3. Dropout의 유무**

In [31]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2599 - accuracy: 0.9258
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1140 - accuracy: 0.9659
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0769 - accuracy: 0.9768
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0570 - accuracy: 0.9826
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0443 - accuracy: 0.9866
313/313 - 1s - loss: 0.0716 - accuracy: 0.9793


[0.07163380831480026, 0.9793000221252441]

In [32]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.5),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3969 - accuracy: 0.8823
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2197 - accuracy: 0.9358
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1829 - accuracy: 0.9453
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1601 - accuracy: 0.9516
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1483 - accuracy: 0.9551
313/313 - 0s - loss: 0.0884 - accuracy: 0.9732


[0.08835285902023315, 0.9732000231742859]

In [33]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2929 - accuracy: 0.9146
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1410 - accuracy: 0.9574
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1077 - accuracy: 0.9675
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0877 - accuracy: 0.9732
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0757 - accuracy: 0.9761
313/313 - 1s - loss: 0.0711 - accuracy: 0.9782


[0.0711447224020958, 0.9782000184059143]




*   Dropout가 없을 때가 약각 학습자체도 빠르고 성능이 좋습니다.
*   Dropout이 0.5보다 0.2가 정확도가 약간 높게 나타납니다.



**4. relu가 있고 없고 dropout**

In [34]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3839 - accuracy: 0.8902
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3155 - accuracy: 0.9115
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3031 - accuracy: 0.9152
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2967 - accuracy: 0.9166
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2910 - accuracy: 0.9186
313/313 - 0s - loss: 0.2993 - accuracy: 0.9178


[0.29928305745124817, 0.9178000092506409]

In [35]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2902 - accuracy: 0.9161
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1370 - accuracy: 0.9586
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1033 - accuracy: 0.9682
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0846 - accuracy: 0.9740
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0728 - accuracy: 0.9771
313/313 - 1s - loss: 0.0716 - accuracy: 0.9784


[0.07159304618835449, 0.9783999919891357]



*   reul가 있을 때 성능과 정확도가 좋게 나옴
*   dropout은 성능게 큰 차이를 주지 못하는 것으로 보임



In [36]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2268 - accuracy: 0.9314
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0992 - accuracy: 0.9695
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0695 - accuracy: 0.9784
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0572 - accuracy: 0.9816
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0445 - accuracy: 0.9852
313/313 - 1s - loss: 0.0941 - accuracy: 0.9739


[0.09412439912557602, 0.9739000201225281]

**5. Keras API ( Sequential vs Functional)**

In [37]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Reshape((28, 28, 1)),
  tf.keras.layers.Conv2D(16, 3, activation='relu'),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [38]:
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1543 - accuracy: 0.9539
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0612 - accuracy: 0.9808
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0417 - accuracy: 0.9865
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0305 - accuracy: 0.9901
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0236 - accuracy: 0.9916
313/313 - 1s - loss: 0.0856 - accuracy: 0.9793


[0.08563069999217987, 0.9793000221252441]

In [39]:
from tensorflow import keras
from tensorflow.keras import layers

In [40]:
inputs = keras.Input(shape=(28, 28))
x = layers.Reshape((28, 28, 1))(inputs)
x = layers.Conv2D(16, 3, activation='relu')(x)
x = layers.Flatten()(x)
x = layers.Dense(128)(x)
x = layers.Dense(10, activation='softmax')(x)
outputs = x

model = keras.Model(inputs, outputs)
model.summary()

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
reshape_1 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 26, 26, 16)        160       
_________________________________________________________________
flatten_13 (Flatten)         (None, 10816)             0         
_________________________________________________________________
dense_34 (Dense)             (None, 128)               1384576   
_________________________________________________________________
dense_35 (Dense)             (None, 10)                1290      
Total params: 1,386,026
Trainable params: 1,386,026
Non-trainable params: 0
____________________________________________

model.summary() 를 통해 모델의 대략적인 구조를 파악해볼 수 있다. 위의 CNN 모델은 1,386,026개의 학습 가능한 파라미터를 가진다.

In [41]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1562 - accuracy: 0.9537
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0632 - accuracy: 0.9803
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0415 - accuracy: 0.9870
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0295 - accuracy: 0.9904
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0207 - accuracy: 0.9936
313/313 - 1s - loss: 0.0846 - accuracy: 0.9789


[0.08459612727165222, 0.9789000153541565]

Keras 모델을 만드는데는 sequential 방법과 functional 방법이 있다.

먼저, sequential 방법은 층층이(layer-by-layer) 쌓아 올릴 수 있게 하는 방법으로 대부분 문제를 해결할 수 있다. 하지만, 이 방법은 layer를 공유하는 구조나, 다중 입력/출력을 사용하지는 못한다는 한계가 있다.

대신, functional 방법을 사용하면, layer가 앞/뒤 layer에만 연결된 구조뿐 아니라 훨씬 더 자유자재로 그 구조를 정의하여 사용할 수 있게된다. layer들을 어떤 layer에든지 연결해서 사용할 수 있는 것이다. 그래서 siamese 네트워크와 residual 네트워크 같은 복잡한 구조의 네트워크를 만들 수 있게 된다


**5. Pooling의 사용**

In [22]:
inputs = keras.Input(shape=(28, 28))
x = layers.Reshape((28, 28, 1))(inputs)
x = layers.Conv2D(16, 3, activation='relu')(x)
x = layers.MaxPooling2D(2)(x)
x = layers.Flatten()(x)
x = layers.Dense(128)(x)
x = layers.Dense(10, activation='softmax')(x)
outputs = x

model = keras.Model(inputs, outputs)
model.summary()

Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
reshape_3 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 26, 26, 16)        160       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 16)        0         
_________________________________________________________________
flatten_13 (Flatten)         (None, 2704)              0         
_________________________________________________________________
dense_32 (Dense)             (None, 128)               346240    
_________________________________________________________________
dense_33 (Dense)             (None, 10)                1290

model.summary() 를 통해 모델의 대략적인 구조를 파악해볼 수 있다. 위의 CNN 모델은 1,386,026개의 학습 가능한 파라미터를 가진다.

In [23]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 36s 604us/sample - loss: 0.1562 - accuracy: 0.9533
Epoch 2/5
60000/60000 [==============================] - 34s 574us/sample - loss: 0.0677 - accuracy: 0.9794
Epoch 3/5
60000/60000 [==============================] - 34s 572us/sample - loss: 0.0517 - accuracy: 0.9842
Epoch 4/5
60000/60000 [==============================] - 34s 572us/sample - loss: 0.0407 - accuracy: 0.9875
Epoch 5/5
60000/60000 [==============================] - 34s 571us/sample - loss: 0.0331 - accuracy: 0.9895
10000/1 - 2s - loss: 0.0316 - accuracy: 0.9828


[0.05949438586093602, 0.9828]

**7. 조금더 깊은 layer**

In [28]:
inputs = keras.Input(shape=(28, 28))
x = layers.Reshape((28, 28, 1))(inputs)
x = layers.Conv2D(16, 3, activation='relu')(x)
x = layers.MaxPooling2D(2)(x)
x = layers.Conv2D(32, 3, activation='relu')(x)
x = layers.MaxPooling2D(2)(x)
x = layers.Flatten()(x)
x = layers.Dense(128)(x)
x = layers.Dense(10, activation='softmax')(x)
outputs = x

model = keras.Model(inputs, outputs)
model.summary()

Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_3 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
reshape_4 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 26, 26, 16)        160       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 16)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 11, 11, 32)        4640      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 32)          0         
_________________________________________________________________
flatten_14 (Flatten)         (None, 800)               0   

In [29]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 42s 706us/sample - loss: 0.1445 - accuracy: 0.9556
Epoch 2/5
60000/60000 [==============================] - 42s 701us/sample - loss: 0.0533 - accuracy: 0.9835
Epoch 3/5
60000/60000 [==============================] - 42s 694us/sample - loss: 0.0394 - accuracy: 0.9873
Epoch 4/5
60000/60000 [==============================] - 41s 690us/sample - loss: 0.0317 - accuracy: 0.9899
Epoch 5/5
60000/60000 [==============================] - 42s 694us/sample - loss: 0.0263 - accuracy: 0.9916
10000/1 - 3s - loss: 0.0176 - accuracy: 0.9884


[0.03441185582805774, 0.9884]

결과 Pooling layer는 이미지의 크기를 작게 만듦으로써 정보의 손실을 가져온다고 볼수도 있다. 하지만 Flatten layer가 있는 모델의 구조상 pooling layer를 통해서 전체 모델의 parameter 양은 되려 줄어든다.

**8. Padding의 유무**

In [30]:
inputs = keras.Input(shape=(28, 28))
x = layers.Reshape((28, 28, 1))(inputs)
x = layers.Conv2D(16, 3, activation='relu',padding="same")(x)
x = layers.MaxPooling2D(2)(x)
x = layers.Conv2D(32, 3, activation='relu',padding="same")(x)
x = layers.MaxPooling2D(2)(x)
x = layers.Flatten()(x)
x = layers.Dense(128)(x)
x = layers.Dense(10, activation='softmax')(x)
outputs = x

model = keras.Model(inputs, outputs)
model.summary()

Model: "model_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_4 (InputLayer)         [(None, 28, 28)]          0         
_________________________________________________________________
reshape_5 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 28, 28, 16)        160       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 14, 14, 16)        0         
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 14, 14, 32)        4640      
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 7, 7, 32)          0         
_________________________________________________________________
flatten_15 (Flatten)         (None, 1568)              0   

In [31]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 52s 867us/sample - loss: 0.1341 - accuracy: 0.9585
Epoch 2/5
60000/60000 [==============================] - 51s 852us/sample - loss: 0.0487 - accuracy: 0.9844
Epoch 3/5
60000/60000 [==============================] - 51s 856us/sample - loss: 0.0382 - accuracy: 0.9876
Epoch 4/5
60000/60000 [==============================] - 51s 858us/sample - loss: 0.0288 - accuracy: 0.9908
Epoch 5/5
60000/60000 [==============================] - 52s 861us/sample - loss: 0.0254 - accuracy: 0.9919
10000/1 - 3s - loss: 0.0187 - accuracy: 0.9893


[0.03719957101760156, 0.9893]

학습 및 검증 결과 검증 정확도 98.11%로 간단한 모델 치고는 나쁘지 않은 성능이 나왔다.